java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java

```diff
@@ -126,141 +115,111 @@ public String toString() {
+      return combineTransitively(a, b);
+    }
+    return null;
   }
 
-  /**
-   * Create a new relation, if any, that is a transitive combination of the receiver with the supplied relation.
-   * Schema aSb & bRc => aTc
-   * where aRb is the relation described by the receiver, bSc that of the supplied relation, aTc is the returned relation.
-   * 
-   * @param relation another SymbolicValueRelation that can be combined transitively with the receiver.
-   * @return a SymbolicValueRelation or null if the receiver and the supplied relation cannot be combined
-   */
   @CheckForNull
-  private BinaryRelation combineOrdered(BinaryRelation relation) {
-    Preconditions.checkArgument(leftOp.equals(relation.rightOp), "Transitive condition not matched!");
-    if (rightOp.equals(relation.leftOp)) {
-      return conjunction(relation.symmetric());
+  private static BinaryRelation simplify(BinaryRelation rel1, BinaryRelation rel2) {
+    // a >= b && b >= a -> a == b
+    if (rel1.kind == GREATER_THAN_OR_EQUAL && rel2.kind == GREATER_THAN_OR_EQUAL
+      && rel1.hasSameOperandsAs(rel2) && !rel1.equals(rel2)) {
+      return binaryRelation(EQUAL, rel1.leftOp, rel1.rightOp);
     }
-    return combinedAfter(relation);
+    return null;
   }
 
-  /**
-   * Returns a new relation resulting of the conjunction of the receiver with the supplied relation.
-   * Schema aRb & aSb => aTb
-   * where aRb is the relation described by the receiver, bSc that of the supplied relation, aTc is the returned relation.
-   * 
-   * @param relation another relation bearing on the same operands as those of the receiver
-   * @return the resulting relation or null if the conjunction of both relations does not bring any new information
-   */
+  private boolean hasSameOperand() {
+    return leftOp.equals(rightOp);
+  }
 
-  @CheckForNull
-  protected BinaryRelation conjunction(BinaryRelation relation) {
-    Preconditions.checkArgument(leftOp.equals(relation.leftOp) && rightOp.equals(relation.rightOp), "Conjunction condition not matched!");
-    return null;
+  private boolean hasOperand(SymbolicValue operand) {
+    return leftOp.equals(operand) || rightOp.equals(operand);
+  }
+
+  private boolean hasSameOperandsAs(BinaryRelation other) {
+    return (leftOp.equals(other.leftOp) && rightOp.equals(other.rightOp))
+      || (leftOp.equals(other.rightOp) && rightOp.equals(other.leftOp));
+  }
+
+  @VisibleForTesting
+  SymbolicValue differentOperand(BinaryRelation other) {
+    Preconditions.checkState(inTransitiveRelationship(other), "%s is not in transitive relationship with %s", this, other);
+    return other.hasOperand(leftOp) ? rightOp : leftOp;
+  }
+
+  @VisibleForTesting
+  SymbolicValue commonOperand(BinaryRelation other) {
+    Preconditions.checkState(inTransitiveRelationship(other));
+    return other.hasOperand(leftOp) ? leftOp : rightOp;
+  }
+
+  @VisibleForTesting
+  boolean inTransitiveRelationship(BinaryRelation other) {
```

Comment:
Naming here : I don't think this current name makes much sense. 
What about  : canGenerateATransitiveRelation ?